client.go

```diff
@@ -0,0 +1,382 @@
+	client := &Client{
+		name:                 name,
+		clusterConfig:        &config.ClusterConfig,
+		serializationService: serializationService,
+		eventDispatcher:      event.NewDispatchService(),
+		userEventDispatcher:  event.NewDispatchService(),
+		logger:               clientLogger,
+		refIDGen:             iproxy.NewReferenceIDGenerator(),
+	}
+	client.subscribeUserEvents()
+	client.createComponents(&config)
+	return client, nil
+}
+
+// Name returns client's name
+// Use ConfigBuilder.SetName to set the client name.
+// If not set manually, an automatic name is used.
+func (c *Client) Name() string {
+	return c.name
+}
+
+// GetMap returns a distributed map instance.
+func (c *Client) GetMap(name string) (*Map, error) {
+	return c.GetMapContext(context.Background(), name)
+}
+
+// GetMapContext returns a distributed map instance.
+func (c *Client) GetMapContext(ctx context.Context, name string) (*Map, error) {
+	if atomic.LoadInt32(&c.state) != ready {
+		return nil, ErrClientNotReady
+	}
+	if ctx == nil {
+		return nil, ErrContextIsNil
+	}
+	return c.proxyManager.getMapWithContext(ctx, name)
+}
+
+// GetReplicatedMap returns a replicated map instance.
+func (c *Client) GetReplicatedMap(name string) (*ReplicatedMap, error) {
+	if atomic.LoadInt32(&c.state) != ready {
+		return nil, ErrClientNotReady
+	}
+	return c.GetReplicatedMapContext(context.Background(), name)
+}
+
+// GetReplicatedMapContext returns a replicated map instance.
+func (c *Client) GetReplicatedMapContext(ctx context.Context, name string) (*ReplicatedMap, error) {
+	if atomic.LoadInt32(&c.state) != ready {
+		return nil, ErrClientNotReady
+	}
+	if ctx == nil {
+		return nil, ErrContextIsNil
+	}
+	return c.proxyManager.getReplicatedMapWithContext(ctx, name)
+}
+
+// Start connects the client to the cluster.
+func (c *Client) start() error {
+	if !atomic.CompareAndSwapInt32(&c.state, created, starting) {
+		return ErrClientCannotStart
+	}
+	// TODO: Recover from panics and return as error
+	c.eventDispatcher.Publish(newLifecycleStateChanged(LifecycleStateStarting))
+	clusterServiceStartCh := c.clusterService.Start(c.clusterConfig.SmartRouting)
+	c.partitionService.Start()
+	if err := c.connectionManager.Start(); err != nil {
+		return err
+	}
+	<-clusterServiceStartCh
+	atomic.StoreInt32(&c.state, ready)
+	c.eventDispatcher.Publish(newLifecycleStateChanged(LifecycleStateStarted))
+	return nil
+}
+
+// Shutdown disconnects the client from the cluster.
+func (c *Client) Shutdown() error {
+	if !atomic.CompareAndSwapInt32(&c.state, ready, stopping) {
+		return ErrClientNotReady
+	}
+	c.eventDispatcher.Publish(newLifecycleStateChanged(LifecycleStateShuttingDown))
+	c.invocationService.Stop()
+	c.clusterService.Stop()
+	c.partitionService.Stop()
+	<-c.connectionManager.Stop()
+	atomic.StoreInt32(&c.state, stopped)
+	c.eventDispatcher.Publish(newLifecycleStateChanged(LifecycleStateShutDown))
+	// wait for the shut down ebent to be dispatched
+	time.Sleep(10 * time.Millisecond)
```

Comment:
We should not need this sleep. 
anything pushed to a channel will be consumed before close anyway.

Reply:
I just wrote a test noticed that it is not guaranteed that all messages will be consumed. Still we should have a better way than the sleep. This can not guarantee that even will be dispathced

Reply:
The following statement is true for channels
`anything pushed to a channel will be consumed before close`
I think what we should do is to consume the eventCh when doneCh is closed, until eventCh starts to return false.  
So adding something the following range loop to where we wait for doneCh should do the trick if I am not mistaken
```
case <-s.doneCh:
			for event := range s.eventCh {
				s.dispatch(event)
			}		
			return
```